### PR TITLE
dnstype, ipnlocal: dnstype.Resolver.UseWithExitNode controls exit node DNS behavior

### PR DESCRIPTION
**Pull request by mzbenami**
Updates #8237 

Fixes tailscale/corp#30906
Fixes tailscale/corp#30907

- dnstype.Resolver  adds a boolean UseWithExitNode that controls whether the resolver should be used in tailscale exit node contexts (not wireguard exit nodes)
- If UseWithExitNode resolvers are found, they are installed as the global resolvers
- If no UseWithExitNode resolvers are found, the exit node resolver continues to be installed as the global resolver
- Split DNS Routes referencing UseWithExitNode resolvers are also installed.

I wanted to get this PR up, but I am in parallel bringing these changes into corp, making sure it compiles, and doing some light local controller testing.

This PR also adds some test coverage split DNS routes with zero-length resolver lists, independent of UseWithExitNode routes.

A former version of this PR as of d12c7a4 created a new dns type in tailcfg that embedded the dnstype.Resolver, but after some re-evaluation, we decided to back to the original plan. I'm happy to chat more about that decision out of band.

### CODE REVIEW

File: tailcfg/tailcfg.go [hunk 1]
```diff
@@ -1698,10 +1699,17 @@ var FilterAllowAll = []FilterRule{
 	},
 }
 
+// DNSResolver embeds dnstype.Resolver and stores
+// additional configuration.
+type DNSResolver struct {
+	dnstype.Resolver `json:",omitempty"`
+	UseWithExitNode  bool `json:",omitempty"`
```
Comment (sailorfrag) [inline tailcfg/tailcfg.go]: Needs docstring, one suggestion:
```
    // UseWithExitNode designates that this resolver should continue to be used when an
    // exit node is in use. Normally, DNS resolution is delegated to the exit node but
    // there are situations where it is preferable to still use a Split DNS server and/or
    // global DNS server instead of the exit node. 
```

File: tailcfg/tailcfg.go [hunk 2]
```diff
@@ -1698,10 +1699,17 @@ var FilterAllowAll = []FilterRule{
 	},
 }
 
+// DNSResolver embeds dnstype.Resolver and stores
+// additional configuration.
```
Comment (sailorfrag) [inline tailcfg/tailcfg.go]: This says what it is (which can be seen below), but not really what it's used for. Maybe something like
```
// DNSResolver describes a single DNS resolver and any special handling needed when
// using that resolver.
```

I guess that's not significantly more informative but I feel like it reads better?

**Comment by mzbenami** [comment]
Converted this back to draft, after seeing the effect on corp

File: client/tailscale/apitype/controltype.go [hunk 3]
```diff
@@ -16,4 +16,5 @@ type DNSConfig struct {
 type DNSResolver struct {
 	Addr                string   `json:"addr"`
 	BootstrapResolution []string `json:"bootstrapResolution,omitempty"`
+	UseWithExitNode     bool     `json:"useWithExitNode"`
```
Comment (bradfitz) [inline client/tailscale/apitype/controltype.go]: omitempty? 

and add docs while you're here? Or at least a reference/URL to where the docs are elsewhere?

File: ipn/ipnlocal/node_backend.go [hunk 4]
```diff
@@ -700,10 +736,42 @@ func dnsConfigForNetmap(nm *netmap.NetworkMap, peers map[tailcfg.NodeID]tailcfg.
...
+			// can authoritatively answer "name not exists" for when the
+			// control plane also sends this explicit but empty route
+			// making it as something we handle.
+			//
+			// While we're already populating it, might as well size the
+			// slice appropriately.
+			// Per #9498 the exact requirements of nil vs empty slice remain
+			// unclear, this is a haunted graveyard to be resolved.
+			dcfg.Routes[fqdn] = make([]*dnstype.Resolver, 0, len(resolvers))
+			dcfg.Routes[fqdn] = append(dcfg.Routes[fqdn], resolvers...)
```
Comment (bradfitz) [inline ipn/ipnlocal/node_backend.go]: kinda weird make + append here?

Is this just a slices.Clone?

File: ipn/ipnlocal/node_backend.go [hunk 5]
```diff
@@ -578,6 +578,42 @@ func (nb *nodeBackend) doShutdown(cause error) {
 	nb.eventClient.Close()
 }
 
+// useWithExitNodeResolvers filters out resolvers so the ones that remain
+// are all the ones marked for use with exit nodes.
+func useWithExitNodeResolvers(resolvers []*dnstype.Resolver) []*dnstype.Resolver {
+	filtered := make([]*dnstype.Resolver, 0, len(resolvers))
```
Comment (bradfitz) [inline ipn/ipnlocal/node_backend.go]: not sure this pre-sizing will be worth it. It'll probably be too big usually.

I'd just start with nil and use append and let append doubling do its job

File: ipn/ipnlocal/node_backend.go [hunk 6]
```diff
@@ -578,6 +578,42 @@ func (nb *nodeBackend) doShutdown(cause error) {
...
+		}
+	}
+	return filtered
+}
+
+// useWithExitNodeRoutes filters out routes so the ones that remain
+// are either zero-length resolver lists, or lists containing only
+// resolvers marked for use with exit nodes.
+func useWithExitNodeRoutes(routes map[string][]*dnstype.Resolver) map[string][]*dnstype.Resolver {
+	filtered := make(map[string][]*dnstype.Resolver)
```
Comment (bradfitz) [inline ipn/ipnlocal/node_backend.go]: could lazily allocate this, right? and return nil in the common case? most users won't have anything here.

File: client/tailscale/apitype/controltype.go (hunk 3 above)
Comment (mzbenami) [reply client/tailscale/apitype/controltype.go]: bfb4dc42d

Interestingly enough, this seems to be missing from public API docs. https://tailscale.com/api#tag/dns
Not sure if that's intentional or should be a ticket.

I wasn't sure what level of documentation was appropriate, given this struct has high overlap with the DNSConfig in tailcfg, and the existing API types vary in documentation style. So I did field-level comments that are sometimes shorter than their tailcfg counterparts.

I also took the opportunity to move Nameservers to the bottom of both type definitions, as it is used in legacy contexts, as I understand.

File: ipn/ipnlocal/node_backend.go (hunk 4 above)
Comment (mzbenami) [reply ipn/ipnlocal/node_backend.go]: bfb4dc42d
Yup, this was a copy/paste as part of a small refactor, but agree it is just a slices.Clone.

File: ipn/ipnlocal/node_backend.go (hunk 5 above)
Comment (mzbenami) [reply ipn/ipnlocal/node_backend.go]: bfb4dc42d
Good call.

File: ipn/ipnlocal/node_backend.go (hunk 6 above)
Comment (mzbenami) [reply ipn/ipnlocal/node_backend.go]: bfb4dc42d
Good call.

**Comment by mzbenami** [comment]
@bradfitz 
> Squash the commits and make the commit message (including title) match http://go/commit style?

I did this, and all review feedback changes are in a single commit rebased on a recent origin/main.

I'm curious if this was the right move, or if we prefer leaving a commit trail as feedback is addressed, and then doing a manual rebase and force push at the very end?

In any case, leaving a commit trail, and then hitting "Squash and merge" is not what we want, right?

File: ipn/ipnlocal/node_backend.go (hunk 4 above)
Comment (mzbenami) [reply ipn/ipnlocal/node_backend.go]: I should actually call out, that the old behavior was that a nil resolvers slice in the map would get converted to an allocated empty slice. Now, with slices.Clone, nil in means nil out. If `dcfg.Routes` is being serialized to another representation, that might actually make a difference somewhere.

File: client/tailscale/apitype/controltype.go [hunk 7]
```diff
@@ -3,17 +3,45 @@
...
+// used in /tailnet/{tailnet}/dns/config.
 type DNSConfig struct {
-	Resolvers          []DNSResolver            `json:"resolvers"`
-	FallbackResolvers  []DNSResolver            `json:"fallbackResolvers"`
-	Routes             map[string][]DNSResolver `json:"routes"`
-	Domains            []string                 `json:"domains"`
-	Nameservers        []string                 `json:"nameservers"`
-	Proxied            bool                     `json:"proxied"`
-	TempCorpIssue13969 string                   `json:"TempCorpIssue13969,omitempty"`
+	// Resolvers are the DNS resolvers to use.
```
Comment (sailorfrag) [inline client/tailscale/apitype/controltype.go]: global DNS resolvers (overriding the local OS configuration)

File: client/tailscale/apitype/controltype.go [hunk 8]
```diff
@@ -3,17 +3,45 @@
...
+	Resolvers []DNSResolver `json:"resolvers"`
+
+	// FallbackResolvers are used in special split DNS configurations.
```
Comment (sailorfrag) [inline client/tailscale/apitype/controltype.go]: FallbackResolvers are used as global resolvers when the client is unable to determine the OS's preferred DNS servers

File: client/tailscale/apitype/controltype.go [hunk 9]
```diff
@@ -3,17 +3,45 @@
...
+	// See https://github.com/tailscale/tailscale/issues/1743.
```
Comment (sailorfrag) [inline client/tailscale/apitype/controltype.go]: Full URLs are probably fine, but I think there's a preference for `#xxx` when the issue is in the same repo or `tailscale/corp#xxx` for another repo.

File: client/tailscale/apitype/controltype.go [hunk 10]
```diff
@@ -3,17 +3,45 @@
...
+
+	// Proxied means MagicDNS is enabled.
+	Proxied bool `json:"proxied"`
+
+	// TempCorpIssue13969 is from an internal hack day prototype,
+	// See https://github.com/tailscale/corp/issues/13969.
+	TempCorpIssue13969 string `json:"TempCorpIssue13969,omitempty"`
+
+	// Nameservers are the IP addresses of the nameservers to use, and
+	// are used in some legacy places instead of Resolvers.
```
Comment (sailorfrag) [inline client/tailscale/apitype/controltype.go]: Nameservers are the IP addresses of global nameservers to use. This is a deprecated format but may still be found in tailnets that were configured a long time ago. When making updates, set Resolvers and leave Nameservers empty.

File: client/tailscale/apitype/controltype.go [hunk 11]
```diff
@@ -3,17 +3,45 @@
...
+	Nameservers []string `json:"nameservers"`
 }
 
+// DNSResolver is a DNS resolver in a DNS configuration.
 type DNSResolver struct {
-	Addr                string   `json:"addr"`
+	// Addr is the address of the DNS resolver.
```
Comment (sailorfrag) [inline client/tailscale/apitype/controltype.go:40]: It is usually an IP address or a DoH URL. See dnstype.Resolver.Addr for full details.

File: tailcfg/tailcfg.go [hunk 12]
```diff
@@ -1767,6 +1762,12 @@ type DNSConfig struct {
 	// It contains a user inputed URL that should have a list of domains to be blocked.
 	// See https://github.com/tailscale/corp/issues/13969.
 	TempCorpIssue13969 string `json:",omitempty"`
+
+	// The following fields are only set and used by
+	// MapRequest.Version >=9 and <14.
+
+	// Nameservers are the IP addresses of the nameservers to use.
+	Nameservers []netip.Addr `json:",omitempty"`
```
Comment (sailorfrag) [inline tailcfg/tailcfg.go]: I think rather than moving this down, incorporate the version range in the comment:

```suggestion
	// Nameservers are the IP addresses of the global nameservers to use.
	//
	// Deprecated: this is only set and used by MapRequest.Version >=9 and <14. Use Resolvers instead.
	Nameservers []netip.Addr `json:",omitempty"`
```

File: ipn/ipnlocal/node_backend.go [hunk 13]
```diff
@@ -578,6 +578,48 @@ func (nb *nodeBackend) doShutdown(cause error) {
...
+func useWithExitNodeRoutes(routes map[string][]*dnstype.Resolver) map[string][]*dnstype.Resolver {
+	var filtered map[string][]*dnstype.Resolver
+	for suffix, resolvers := range routes {
+		// Suffixes with no resolvers represent a valid configuration,
+		// and should persist regardless of exit node considerations.
+		if len(resolvers) == 0 {
+			if filtered == nil {
+				filtered = map[string][]*dnstype.Resolver{}
+			}
+			filtered[suffix] = make([]*dnstype.Resolver, 0)
```
Comment (sailorfrag) [inline ipn/ipnlocal/node_backend.go]: We have a helper to eliminate the `nil` check, since it's so common.
```suggestion
			mak.Set(&filtered, suffix, make([]*dnstype.Resolver, 0))
```

File: ipn/ipnlocal/node_backend.go [hunk 14]
```diff
@@ -700,10 +742,41 @@ func dnsConfigForNetmap(nm *netmap.NetworkMap, peers map[tailcfg.NodeID]tailcfg.
...
+			}
+
+			// Create map entry even if len(resolvers) == 0; Issue 2706.
+			// This lets the control plane send ExtraRecords for which we
+			// can authoritatively answer "name not exists" for when the
+			// control plane also sends this explicit but empty route
+			// making it as something we handle.
+			//
+			// While we're already populating it, might as well size the
+			// slice appropriately.
```
Comment (sailorfrag) [inline ipn/ipnlocal/node_backend.go]: I think this part is no longer relevant?

**Review by sailorfrag — APPROVED** [review]
Some comments to fix up and other minor things, but otherwise looks good.
